Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/impl/apmlog/LogFilterControllerWithAPMLog.java

```diff
@@ -24,55 +24,193 @@
+            parameters.put("value", data);
+        } else if (filterType == FilterType.ATTRIBUTE_ARC_DURATION) {
+            source = (String) param.get("source");
+            target = (String) param.get("target");
+            if (!logData.hasSufficientDurationVariant(mainAttribute, source, target)) {
+                Messagebox.show("Unable to filter on arc duration as there's only one value.",
+                        "Filter error", Messagebox.OK, Messagebox.ERROR);
+                return;
+            }
+            parameters.put("key", mainAttribute);
+            parameters.put("from", source);
+            parameters.put("to", target);
+        } else {
+            return;
+        }
+        Clients.showBusy("Launch Filter Dialog ...");
+
+        LogFilterRule rule = getLastMatchedRuleWithValues(filterType, parameters,
+                (List<LogFilterRule>) logData.getCurrentFilterCriteria());
+
+        EditorOption option = rule != null ? new EditorOption(filterType, rule) :
+                new EditorOption(filterType, parameters);
+
+        LogFilterRequest lfr = getRequestWithOption(option);
+        parent.getLogFilterPlugin().execute(lfr);
+        Clients.clearBusy();
+    }
+
+    private LogFilterRequest getRequestWithOption(EditorOption option) {
+        return new LogFilterRequest(this, parent.getSourceLogId(), parent.getTitle(),
+                logData.getOriginalAPMLog(), (List<LogFilterRule>) logData.getCurrentFilterCriteria(), option);
+    }
+
+    private LogFilterRequest getDefaultRequest() {
+        return new LogFilterRequest(this, parent.getSourceLogId(), parent.getTitle(),
+                logData.getOriginalAPMLog(), (List<LogFilterRule>) logData.getCurrentFilterCriteria());
+    }
+
+    private FilterType getFilterType(String payload) {
+        switch (payload) {
+            case "CaseTabID": return FilterType.CASE_ID;
+            case "CaseTabVariant": return FilterType.CASE_VARIANT;
+            case "EventTabAttribute": return FilterType.EVENT_EVENT_ATTRIBUTE;
+            case "CaseTabAttribute": return FilterType.CASE_EVENT_ATTRIBUTE;
+            case "CaseTabPerformance": return FilterType.DURATION;
+            case "CaseTabTimeframe": return FilterType.CASE_TIME;
+            case "CASE_SECTION_ATTRIBUTE_COMBINATION": return FilterType.CASE_SECTION_ATTRIBUTE_COMBINATION;
+            case "EVENT_ATTRIBUTE_DURATION": return FilterType.EVENT_ATTRIBUTE_DURATION;
+            case "ATTRIBUTE_ARC_DURATION": return FilterType.ATTRIBUTE_ARC_DURATION;
+        }
+        return FilterType.UNKNOWN;
+    }
+
+    private LogFilterRule getLastMatchedRule(FilterType filterType, List<LogFilterRule> criteria) {
+        if (criteria == null || criteria.isEmpty()) return null;
+
+        for (int i = criteria.size() - 1; i >= 0; i--) {
+            if (criteria.get(i).getFilterType() == filterType) {
+                return criteria.get(i);
+            }
+        }
+
+        return null;
```

Comment:
This could be reduced to a one-liner using streams:
```
return criteria.stream().filter(criterion -> criterion.getFilterType() == filterType).findFirst().orElse(null);
```